compiler/noirc_evaluator/src/ssa/ir/instruction/binary.rs

```diff
@@ -116,13 +122,36 @@ pub(crate) fn eval_constant_binary_op(
             // Thus, we do not evaluate the division in this method, as we want to avoid triggering a panic,
             // and the operation should be handled by ACIR generation.
             if matches!(operator, BinaryOp::Div | BinaryOp::Mod) && rhs == 0 {
-                return None;
+                return Some(Err("attempt to divide by zero".to_string()));
             }
-            let result = function(lhs, rhs)?;
+
+            let Some(result) = function(lhs, rhs) else {
+                if let BinaryOp::Shl = operator {
+                    return None;
+                }
+
+                if let BinaryOp::Shr = operator {
+                    return Some(Ok((FieldElement::zero(), operand_type)));
+                }
+
+                let op = binary_op_function_name(operator);
+                return Some(Err(format!("attempt to {op} with overflow")));
+            };
+
             // Check for overflow
             if result != 0 && result.ilog2() >= bit_size {
-                return None;
+                if let BinaryOp::Shl = operator {
+                    return None;
+                }
+
+                if let BinaryOp::Shr = operator {
+                    return Some(Ok((FieldElement::zero(), operand_type)));
```

Comment:
This doesn't look right: if `result != 0` then we shouldn't return `FieldElement::zero()`. It was more like an overflowing multiplication, ending up with more bit size than what fits the type, so this is an `Err` I think.

Reply:
In this case it's a shift-right, and I _think_ that when that "overflows" we return zero... but I might be wrong. Shifting has been very confusing lately 😅